File: python/easy/145. Binary Tree Postorder Traversal.py
# Given the root of a binary tree, return the postorder traversal of its nodes' values.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def postorderTraversal(self, root: TreeNode) -> list[int]:
        tree_values = list()
        if root is None:
            return tree_values
        tree_values += self.postorderTraversal(root.left)
        tree_values += self.postorderTraversal(root.right)
        tree_values.append(root.val)

        return tree_values

    def postorderTraversal2(self, root: TreeNode) -> list[int]:
        if not root:
            return []

        stack = [root]
        result = []

        while stack:
            node = stack.pop()
            result.append(node.val)
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)

        return result[::-1]

File: go/easy/145_Binary_Tree_Postorder_Traversal.go
package main

import "fmt"

//Given the root of a binary tree, return the postorder traversal of its nodes' values.
//
//Example 1:
//Input: root = [1,null,2,3]
//Output: [3,2,1]
//
//Example 2:
//Input: root = []
//Output: []
//
//Example 3:
//Input: root = [1]
//Output: [1]

//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

func postorderHelper(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}
	postorderHelper(root.Left, res)
	postorderHelper(root.Right, res)
	*res = append(*res, root.Val)
}

func postorderTraversal(root *TreeNode) []int {
	var res []int
	postorderHelper(root, &res)
	return res
}

func main() {
	tree := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 0,
		},
		Right: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 3,
			},
		},
	}
	fmt.Println(postorderTraversal(tree))
}
